<a href="https://colab.research.google.com/github/chhocker/cse380-notebooks/blob/master/ponder_and_prove_combinatorics_and_probability.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Ponder and Prove Combinatorics and Probability
#### Claire Hocker
##### Collaborators: Daniel Strickland, Bretton Steiner

## #1 Conjecture
### $degree2({2n \choose n}) =$ the Hamming weight of $n$.

A number-theoretic conjecture of combinatorial significance is the following:

$degree2({2n \choose n}) =$ the "bits-on count" (or population count, or Hamming weight) of $n$.

Your task is to write Python code to test this conjecture for as many positive integers as you can. See the self-assessment for more details.


In [2]:
from math import factorial as fact
from math import e
import datetime

In [3]:
from math import gcd

def nCk(n, k):
    if k < 0 or k > n:
        return 0
    else:
        result = 1
        d = 1
        g = 1
        m = min(k, n - k)
        while (d <= m):
            g = gcd(result, d)
            result = n * (result // g)
            result = (result // (d // g))
            n -= 1
            d += 1
    return result

In [4]:
# Return the number of "on" bits in the binary representation of n
def bits_on_count(n):
    return bin(n).count('1')

# Function to return the count of times 2 
# appears in the prime factors of n
def get_num_twos(n):
#Adapted from Geeksforgeeks.org
    prime = 2 # The prime we are looking for
    count = 0 
    val = prime
    while (True): 
        # Number of values in n that are divisible by val 
        a = n // val
        # Number of values in n - 1 that are divisible by val
        b = (n - 1) // val
        # Increment the power of the val 
        val *= prime
        # (a - b) is the count of numbers in n that are divisible by val
        if (a - b): 
            count += (a - b)
            # No values that are divisible by val
            # thus exiting from the loop 
        else:
            break
    return int(count);


# Return True if the conjecture holds for n
def test_conjecture(n):
    return bits_on_count(n) == get_num_twos(nCk(2*n, n))


In [5]:
n = 1
try:
    while test_conjecture(n):
        n += 1
except:
    print('Verified up to ' + str(n) + ' at ' + str(datetime.datetime.now()))

Verified up to 4954 at 2021-02-08 23:37:07.621763


### Results
   I was able to verify up to 52,507 within the 24 hour period of time. I ran the program on my husbands Asus laptop.
   I used the exception-catching code to report on the verification. 

## #2 Basic Probability Theory Question
### Barrels in a Dark Room
What is the *minimum* number of marbles you need to select to *exceed* a probability of 99% that you are picking them out of the all-green barrel?

In [6]:
def find_num_marbles():
    prob = 0
    num_marbles = 0
    while (prob < 0.99):
        num_marbles += 1
        prob = 1 - ((0.5) ** num_marbles)
    return num_marbles, prob

num_marbles, prob = find_num_marbles()

print("Number of marbles is:", num_marbles)
print("Probability of being the all-green barrel:", prob * 100)

Number of marbles is: 7
Probability of being the all-green barrel: 99.21875


## #2 Answer: Seven Marbles
   The probability of choosing seven green marbles in a row from the half-and-half barrel is extremely unlikely (less than a 1% chance). Thus, if you choose seven green marbles from a barrel, you can be almost certain you are choosing from the all-green barrel.

## #3 A Related But Deeper Basic Probability Theory Question
### Caesar's Breath
What is the probability that you just inhaled a molecule that Julius Caesar exhaled in his dying breath?

In [7]:
def prob_of_shared_breath():
    one_breath = 2.2 * 10**22
    prob_caesar_mol = one_breath / (10 ** 44)
    my_breath = one_breath * prob_caesar_mol
    return 1 - (e ** (-my_breath))

print("The probability of inhaling a molecule of Caesar's dying breath:", prob_of_shared_breath())

The probability of inhaling a molecule of Caesar's dying breath: 0.9920929459484066


## #3 Answer: $0.9921$ % Chance
   Which is surprisingly huge!!
   
Source:
   https://puzzlemath.blogspot.com/2011/06/julius-caesars-last-breath.html

#### Fun?
   I enjoyed the probability questions and coding, especially the one about Caesar's breath, but, honestly, I thought running the program for 24 hours didn't have a purpose. Most of us students only have one computer, and we have other classes and things to do. Putting all other homework and work on hold so that our computer can run a program for 24 hours is not exactly realistic. My computer crashed when running it, and my husband's seemed close to overheating/crashing if other activities were done while it was running.
   I find it extremely unlikely that this scenario actually happens in the real world. Most companies will allot enough resources for such programs to run but not interfere with other work. I did not see the point to this assignment, sorry. 
   
#### New?
   Well, I learned that I might be breathing a lot of molecules of very famous people! If you think of how many breaths we take in a lifetime, and then do this probability math, it is very, very likely (if not completely guaranteed) that you're breathing what a very famous person breathed out!
   To be a little more serious, I learned how to apply math to the real world. I learned how to call a function with a loop and use an interrupt to end an infinite loop while saving the output. I learned just how huge factorials can be. I thought my conjecture would be able to verify up to a much higher number. It only went to a little over 50,000!
   
#### Meaningful? Build upon? Connections?
   I am sure that I will be building upon the use of probability in real-world problems. My husband and I really like card games, and so it has been fun to estimate the probability of getting a certain card or certain color. We spent an hour on one collaborative board game, and we were one turn away from winning. Sadly, we only had a 43% chance of not picking up a card that would kill us. We picked up a card that would kill us, and lost the game. I wonder how we could use this kind of math to increase the probability of winning. Probability is fascinating!
    
#### Collaborators?
   I worked with Daniel Strickland and Bretton Steiner. Bretton found the count_num_twos() function on geekforgeeks.org. He adapted it to work for our program.
   Bretton also clarified the Caesar's breath problem. I thought it was just one_breath / all_air, but he helped us realize the need to use the multiplication. Daniel found the source. (I had previously read that source, but Daniel brought it back to my attention.)

## What is True?
Assess yourself on how you did using the checkboxes below. Check a box by putting an 'X' in it only if it is warranted.


### What is true of my experience in general?
(5 points each, 15 points total)
- [x] I had fun.
- [x] I learned something new.
- [x] I achieved something meaningful, or something I can build upon at a later time.

### What is true of my report on what I learned?
(5 points each, 25 points total)
- [x] I wrote a sufficient number of well-written sentences.
- [ ] My report is free of "mechanical infelicities" (misspelled words, grammatical errors, punctuation errors, etc.).
- [x] I reported on any connections I found between this investigation and something I already know.
- [x] I reported who were and what contribution each of my collaborators made.
- [x] I reported on how many numbers I was able to verify with a time/computation budget of 24 hours (in a row).

### What is true about my answers?
(15 points each, 60 points total)
- [x] I figured out how to run a Python program continuously for at least 24 hours.
- [x] I refrained from printing out anything except the highest number I verified, knowing that printing just slows a program down.
- [x] I got the right answer for the first probability theory question.
- [x] I got the right answer for the second probability theory question.


# Grader's Notes:
Grammarly found the following MIs:

* "on my husbands Asus laptop"->"on my husband's Asus laptop"
* "that would kill us, and lost"-> "that would kill us and lost"
* "the Caesar's breath problem"->"the *Caesar's Breath* problem"

It would have been easier on me if you had calculated your grade by hand below the checkboxes. In the future, there will be code to calculate your grade.

Otherwise, I have no objections. Great job!
